lib/storage/environment.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/gravitational/gravity/lib/constants"
+	"github.com/gravitational/gravity/lib/defaults"
+
+	teleservices "github.com/gravitational/teleport/lib/services"
+	teleutils "github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+// Environment defines the environment variables resource.
+// It allows to override environment variables on each node in the cluster.
+// There is only a single instance of the resource in a cluster
+type Environment interface {
+	// Resource provides common resource methods
+	teleservices.Resource
+	// CheckAndSetDefaults validates this resource and sets defaults
+	CheckAndSetDefaults() error
+	SetKeyValues(map[string]string)
+	GetKeyValues() map[string]string
+}
+
+// NewEnvironment creates a new instance of the resource
+func NewEnvironment(kvs map[string]string) *EnvironmentV2 {
+	return &EnvironmentV2{
+		Kind:    KindEnvironment,
+		Version: teleservices.V2,
+		Metadata: teleservices.Metadata{
+			Name:      constants.ClusterEnvironmentMap,
+			Namespace: defaults.Namespace,
+		},
+		Spec: EnvironmentSpec{
+			KeyValues: kvs,
+		},
+	}
+}
+
+// EnvironmentV2 describes the environment variable resource
```

Comment:
Why V2 if this is the first time environment has been added as an object??

Reply:
This has to do with the teleport resource schema - i.e. it is the `v2` of the teleport schema.

Reply:
This doesn't make sense to me... we bump all resource versions when the teleport schema get's bumped?